openml/flows/sklearn_converter.py

```diff
@@ -0,0 +1,488 @@
+
+    return rval
+
+
+def _is_estimator(o):
+    return (hasattr(o, 'fit') and hasattr(o, 'get_params') and
+            hasattr(o, 'set_params'))
+
+
+def _is_cross_validator(o):
+    return isinstance(o, sklearn.model_selection.BaseCrossValidator)
+
+
+def flow_to_sklearn(o, **kwargs):
+    # First, we need to check whether the presented object is a json string.
+    # JSON strings are used to encoder parameter values. By passing around
+    # json strings for parameters, we make sure that we can flow_to_sklearn
+    # the parameter values to the correct type.
+    if isinstance(o, six.string_types):
+        try:
+            o = json.loads(o)
+        except JSONDecodeError:
+            pass
+
+    if isinstance(o, dict):
+        if 'oml:name' in o and 'oml:description' in o:
+            # TODO check if this code is actually called
+            rval = _deserialize_model(o, **kwargs)
+
+        # Check if the dict encodes a 'special' object, which could not
+        # easily converted into a string, but rather the information to
+        # re-create the object were stored in a dictionary.
+        elif 'oml-python:serialized_object' in o:
+            serialized_type = o['oml-python:serialized_object']
+            value = o['value']
+            if serialized_type == 'type':
+                rval = deserialize_type(value, **kwargs)
+            elif serialized_type == 'rv_frozen':
+                rval = deserialize_rv_frozen(value, **kwargs)
+            elif serialized_type == 'function':
+                rval = deserialize_function(value, **kwargs)
+            elif serialized_type == 'component_reference':
+                value = flow_to_sklearn(value)
+                step_name = value['step_name']
+                key = value['key']
+                component = flow_to_sklearn(kwargs['components'][key])
+                if step_name is None:
+                    rval = component
+                else:
+                    rval = (step_name, component)
+            elif serialized_type == 'cv_object':
+                rval = _deserialize_cross_validator(value, **kwargs)
+            else:
+                raise ValueError('Cannot flow_to_sklearn %s' % serialized_type)
+
+        else:
+            # Regular dictionary
+            rval = OrderedDict((flow_to_sklearn(key, **kwargs),
+                                flow_to_sklearn(value, **kwargs))
+                               for key, value in o.items())
+    elif isinstance(o, (list, tuple)):
+        rval = [flow_to_sklearn(element, **kwargs) for element in o]
+        if isinstance(o, tuple):
+            rval = tuple(rval)
+    elif isinstance(o, (bool, int, float)):
+        rval = o
+    elif isinstance(o, six.string_types):
+        rval = o
+    elif o is None:
+        rval = None
+    elif isinstance(o, OpenMLFlow):
+        rval = _deserialize_model(o, **kwargs)
+    else:
+        raise TypeError(o)
+    assert o is None or rval is not None
```

Comment:
I assume this is a leftover from when I started to develop this to make sure I know where things fail.